main.go

```diff
@@ -0,0 +1,362 @@
+var ErrInvalidSCPType = errors.New("scp type must be Allow or Deny")
+
+type SCPRun struct {
+	scannerFilename string
+	serviceType     string
+	serviceName     string
+	thresholdLimit  int64
+	usageData       []byte
+	reports         *[]Report
+	permissionSet   map[string]int64
+	scp             SCP
+}
+
+//validateService checks that the correct apply or
+//deny value was supplied.
+func (s *SCPRun) validateService() (bool, error) {
+	if !checkSCPParameter(s.serviceType) {
+		return false, ErrInvalidSCPType
+	}
+	return true, nil
+}
+
+func (s *SCPRun) getUsageData() error {
+	usageData, err := loadScannerFile(s.scannerFilename)
+	if err != nil {
+		return err
+	}
+	s.usageData = usageData
+	return nil
+}
+
+func (s *SCPRun) getReport() error {
+	r, err := generateReport(s.usageData)
+	if err != nil {
+		return err
+	}
+	s.reports = r
+	return nil
+}
+
+func (s *SCPRun) createPermissions() error {
+	type fnEval = func(int64, int64) bool
+	var apiFn fnEval
+
+	switch s.serviceType {
+	case "Allow":
+		apiFn = greaterThan
+	case "Deny":
+		apiFn = lessThan
+	}
+
+	r := *s.reports
+	permissionSet, err := generateList(s.thresholdLimit, &r[0], apiFn)
+	if err != nil {
+		return err
+	}
+	s.permissionSet = permissionSet
+	return nil
+}
+
+func (s *SCPRun) formatServiceName() error {
+	r := *s.reports
+	u := &r[0].Results.Service
+
+	s.serviceName = serviceName(*u)
+	return nil
+}
+
+func (s *SCPRun) createSCP() error {
+	s.scp = generateSCP(s.serviceType, s.serviceName, s.permissionSet)
+	return nil
+}
+
+func (s *SCPRun) saveSCP() error {
+	err := saveSCP(s.scp)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
+
+	conf, output, err := parseFlags(os.Args[0],os.Args[1:])
+	if err == flag.ErrHelp {
+		fmt.Fprintln(os.Stderr, exitFail)
+	}
+
+
+	r := new(conf)
+	if err := run(r); err != nil {
+		fmt.Fprintln(os.Stderr,exitFail)
+	}
+
+	fmt.Println(output)
+}
+
+func new(c *SCPConfig) *SCPRun {
+
+	scpRun := SCPRun{scannerFilename: *c.scannerFilename(),
+		serviceType:    *c.serviceType(),
+		thresholdLimit: *c.thresholdLimit()}
+
+	return &scpRun
+}
+
+//run is an abstraction function that allows
+//us to test codebase.
+func run(executionRun *SCPRun) error {
+
+	_, err := executionRun.validateService()
+	if err != nil {
+		return err
+	}
+
+	err = executionRun.getUsageData()
+
+	if err != nil {
+		return err
+	}
+
+	err = executionRun.getReport()
+	if err != nil {
+		return err
+	}
+
+	err = executionRun.createPermissions()
+
+	if err != nil {
+		return err
+	}
+
+	err = executionRun.formatServiceName()
+
+	if err != nil {
+		return err
+	}
+
+	err = executionRun.createSCP()
+
+	if err != nil {
+		return err
+	}
+
+	err = executionRun.saveSCP()
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//SCPConfig is a struct that will hold the
+//flag values
+type SCPConfig struct {
+	SCPType     string
+	ScannerFile string
+	Threshold   int64
+	args []string
```

Comment:
What is the purpose of this args? Looks like they should be loaded to this struct at creation :thinking: